diff --ruN a/stablehlo/stablehlo/tests/chlo/chlo_legalize_to_stablehlo.mlir b/stablehlo/stablehlo/tests/chlo/chlo_legalize_to_stablehlo.mlir
--- stablehlo/stablehlo/tests/chlo/chlo_legalize_to_stablehlo.mlir
+++ stablehlo/stablehlo/tests/chlo/chlo_legalize_to_stablehlo.mlir
@@ -2141,13 +2141,16 @@
 // CHECK:           %[[DIVIDE_19:.*]] = stablehlo.divide %[[MULTIPLY_9]], %[[SINE_0]] : tensor<f64>
 // CHECK:           %[[SUBTRACT_10:.*]] = stablehlo.subtract %[[SUBTRACT_9]], %[[DIVIDE_19]] : tensor<f64>
 // CHECK:           %[[SELECT_1:.*]] = stablehlo.select %[[COMPARE_0]], %[[SUBTRACT_10]], %[[SUBTRACT_9]] : tensor<i1>, tensor<f64>
-// CHECK:           %[[COMPARE_1:.*]] = stablehlo.compare LE, %[[ARG0]], %[[CONSTANT_2]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[COMPARE_1:.*]] = stablehlo.compare EQ, %[[ARG0]], %[[CONSTANT_2]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[COMPARE_2:.*]] = stablehlo.compare LT, %[[ARG0]], %[[CONSTANT_2]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
 // CHECK:           %[[FLOOR_1:.*]] = stablehlo.floor %[[ARG0]] : tensor<f64>
-// CHECK:           %[[COMPARE_2:.*]] = stablehlo.compare EQ, %[[ARG0]], %[[FLOOR_1]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
-// CHECK:           %[[AND_0:.*]] = stablehlo.and %[[COMPARE_1]], %[[COMPARE_2]] : tensor<i1>
+// CHECK:           %[[COMPARE_3:.*]] = stablehlo.compare EQ, %[[ARG0]], %[[FLOOR_1]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[AND_0:.*]] = stablehlo.and %[[COMPARE_2]], %[[COMPARE_3]] : tensor<i1>
 // CHECK:           %[[CONSTANT_25:.*]] = stablehlo.constant dense<0x7FF8000000000000> : tensor<f64>
 // CHECK:           %[[SELECT_2:.*]] = stablehlo.select %[[AND_0]], %[[CONSTANT_25]], %[[SELECT_1]] : tensor<i1>, tensor<f64>
-// CHECK:           return %[[SELECT_2]] : tensor<f64>
+// CHECK:           %[[CONSTANT_26:.*]] = stablehlo.constant dense<0xFFF0000000000000> : tensor<f64>
+// CHECK:           %[[SELECT_3:.*]] = stablehlo.select %[[COMPARE_1]], %[[CONSTANT_26]], %[[SELECT_2]] : tensor<i1>, tensor<f64>
+// CHECK:           return %[[SELECT_3]] : tensor<f64>
 // CHECK:         }
 func.func @digamma_f64(%arg : tensor<f64>) -> tensor<f64> {
   %1 = chlo.digamma %arg : tensor<f64> -> tensor<f64>
@@ -2254,13 +2257,16 @@
 // CHECK:           %[[DIVIDE_19:.*]] = stablehlo.divide %[[MULTIPLY_9]], %[[SINE_0]] : tensor<f32>
 // CHECK:           %[[SUBTRACT_10:.*]] = stablehlo.subtract %[[SUBTRACT_9]], %[[DIVIDE_19]] : tensor<f32>
 // CHECK:           %[[SELECT_1:.*]] = stablehlo.select %[[COMPARE_0]], %[[SUBTRACT_10]], %[[SUBTRACT_9]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[COMPARE_1:.*]] = stablehlo.compare LE, %[[ARG0]], %[[CONSTANT_2]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_1:.*]] = stablehlo.compare EQ, %[[ARG0]], %[[CONSTANT_2]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_2:.*]] = stablehlo.compare LT, %[[ARG0]], %[[CONSTANT_2]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
 // CHECK:           %[[FLOOR_1:.*]] = stablehlo.floor %[[ARG0]] : tensor<f32>
-// CHECK:           %[[COMPARE_2:.*]] = stablehlo.compare EQ, %[[ARG0]], %[[FLOOR_1]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[AND_0:.*]] = stablehlo.and %[[COMPARE_1]], %[[COMPARE_2]] : tensor<i1>
+// CHECK:           %[[COMPARE_3:.*]] = stablehlo.compare EQ, %[[ARG0]], %[[FLOOR_1]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[AND_0:.*]] = stablehlo.and %[[COMPARE_2]], %[[COMPARE_3]] : tensor<i1>
 // CHECK:           %[[CONSTANT_25:.*]] = stablehlo.constant dense<0x7FC00000> : tensor<f32>
 // CHECK:           %[[SELECT_2:.*]] = stablehlo.select %[[AND_0]], %[[CONSTANT_25]], %[[SELECT_1]] : tensor<i1>, tensor<f32>
-// CHECK:           return %[[SELECT_2]] : tensor<f32>
+// CHECK:           %[[CONSTANT_26:.*]] = stablehlo.constant dense<0xFF800000> : tensor<f32>
+// CHECK:           %[[SELECT_3:.*]] = stablehlo.select %[[COMPARE_1]], %[[CONSTANT_26]], %[[SELECT_2]] : tensor<i1>, tensor<f32>
+// CHECK:           return %[[SELECT_3]] : tensor<f32>
 // CHECK:         }
 func.func @digamma_f32(%arg : tensor<f32>) -> tensor<f32> {
   %1 = chlo.digamma %arg : tensor<f32> -> tensor<f32>
@@ -2368,13 +2374,16 @@
 // CHECK:           %[[DIVIDE_19:.*]] = stablehlo.divide %[[MULTIPLY_9]], %[[SINE_0]] : tensor<f32>
 // CHECK:           %[[SUBTRACT_10:.*]] = stablehlo.subtract %[[SUBTRACT_9]], %[[DIVIDE_19]] : tensor<f32>
 // CHECK:           %[[SELECT_1:.*]] = stablehlo.select %[[COMPARE_0]], %[[SUBTRACT_10]], %[[SUBTRACT_9]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[COMPARE_1:.*]] = stablehlo.compare LE, %[[CONVERT_0]], %[[CONSTANT_2]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_1:.*]] = stablehlo.compare EQ, %[[CONVERT_0]], %[[CONSTANT_2]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_2:.*]] = stablehlo.compare LT, %[[CONVERT_0]], %[[CONSTANT_2]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
 // CHECK:           %[[FLOOR_1:.*]] = stablehlo.floor %[[CONVERT_0]] : tensor<f32>
-// CHECK:           %[[COMPARE_2:.*]] = stablehlo.compare EQ, %[[CONVERT_0]], %[[FLOOR_1]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[AND_0:.*]] = stablehlo.and %[[COMPARE_1]], %[[COMPARE_2]] : tensor<i1>
+// CHECK:           %[[COMPARE_3:.*]] = stablehlo.compare EQ, %[[CONVERT_0]], %[[FLOOR_1]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[AND_0:.*]] = stablehlo.and %[[COMPARE_2]], %[[COMPARE_3]] : tensor<i1>
 // CHECK:           %[[CONSTANT_25:.*]] = stablehlo.constant dense<0x7FC00000> : tensor<f32>
 // CHECK:           %[[SELECT_2:.*]] = stablehlo.select %[[AND_0]], %[[CONSTANT_25]], %[[SELECT_1]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[CONVERT_1:.*]] = stablehlo.convert %[[SELECT_2]] : (tensor<f32>) -> tensor<f16>
+// CHECK:           %[[CONSTANT_26:.*]] = stablehlo.constant dense<0xFF800000> : tensor<f32>
+// CHECK:           %[[SELECT_3:.*]] = stablehlo.select %[[COMPARE_1]], %[[CONSTANT_26]], %[[SELECT_2]] : tensor<i1>, tensor<f32>
+// CHECK:           %[[CONVERT_1:.*]] = stablehlo.convert %[[SELECT_3]] : (tensor<f32>) -> tensor<f16>
 // CHECK:           return %[[CONVERT_1]] : tensor<f16>
 // CHECK:         }
 func.func @digamma_f16(%arg : tensor<f16>) -> tensor<f16> {
@@ -2942,17 +2951,20 @@
 // CHECK:           %[[DIVIDE_32:.*]] = stablehlo.divide %[[MULTIPLY_52]], %[[SINE_1]] : tensor<f32>
 // CHECK:           %[[SUBTRACT_18:.*]] = stablehlo.subtract %[[SUBTRACT_17]], %[[DIVIDE_32]] : tensor<f32>
 // CHECK:           %[[SELECT_12:.*]] = stablehlo.select %[[COMPARE_12]], %[[SUBTRACT_18]], %[[SUBTRACT_17]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[COMPARE_13:.*]] = stablehlo.compare LE, %[[ARG1]], %[[CONSTANT_73]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_13:.*]] = stablehlo.compare EQ, %[[ARG1]], %[[CONSTANT_73]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_14:.*]] = stablehlo.compare LT, %[[ARG1]], %[[CONSTANT_73]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
 // CHECK:           %[[FLOOR_5:.*]] = stablehlo.floor %[[ARG1]] : tensor<f32>
-// CHECK:           %[[COMPARE_14:.*]] = stablehlo.compare EQ, %[[ARG1]], %[[FLOOR_5]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[AND_3:.*]] = stablehlo.and %[[COMPARE_13]], %[[COMPARE_14]] : tensor<i1>
+// CHECK:           %[[COMPARE_15:.*]] = stablehlo.compare EQ, %[[ARG1]], %[[FLOOR_5]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[AND_3:.*]] = stablehlo.and %[[COMPARE_14]], %[[COMPARE_15]] : tensor<i1>
 // CHECK:           %[[CONSTANT_96:.*]] = stablehlo.constant dense<0x7FC00000> : tensor<f32>
 // CHECK:           %[[SELECT_13:.*]] = stablehlo.select %[[AND_3]], %[[CONSTANT_96]], %[[SELECT_12]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[SELECT_14:.*]] = stablehlo.select %[[COMPARE_11]], %[[SELECT_13]], %[[MULTIPLY_42]] : tensor<i1>, tensor<f32>
+// CHECK:           %[[CONSTANT_98:.*]] = stablehlo.constant dense<0xFF800000> : tensor<f32>
+// CHECK:           %[[SELECT_16:.*]] = stablehlo.select %[[COMPARE_13]], %[[CONSTANT_98]], %[[SELECT_13]] : tensor<i1>, tensor<f32>
+// CHECK:           %[[SELECT_14:.*]] = stablehlo.select %[[COMPARE_11]], %[[SELECT_16]], %[[MULTIPLY_42]] : tensor<i1>, tensor<f32>
 // CHECK:           %[[FLOOR_6:.*]] = stablehlo.floor %[[ARG0]] : tensor<f32>
-// CHECK:           %[[COMPARE_15:.*]] = stablehlo.compare NE, %[[ARG0]], %[[FLOOR_6]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[COMPARE_16:.*]] = stablehlo.compare LT, %[[ARG0]], %[[CONSTANT_70]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[OR_0:.*]] = stablehlo.or %[[COMPARE_15]], %[[COMPARE_16]] : tensor<i1>
+// CHECK:           %[[COMPARE_16:.*]] = stablehlo.compare NE, %[[ARG0]], %[[FLOOR_6]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_17:.*]] = stablehlo.compare LT, %[[ARG0]], %[[CONSTANT_70]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[OR_0:.*]] = stablehlo.or %[[COMPARE_16]], %[[COMPARE_17]] : tensor<i1>
 // CHECK:           %[[CONSTANT_97:.*]] = stablehlo.constant dense<0x7FC00000> : tensor<f32>
 // CHECK:           %[[SELECT_15:.*]] = stablehlo.select %[[OR_0]], %[[CONSTANT_97]], %[[SELECT_14]] : tensor<i1>, tensor<f32>
 // CHECK:           return %[[SELECT_15]] : tensor<f32>
@@ -3332,17 +3344,20 @@
 // CHECK:           %[[DIVIDE_32:.*]] = stablehlo.divide %[[MULTIPLY_52]], %[[SINE_1]] : tensor<f64>
 // CHECK:           %[[SUBTRACT_18:.*]] = stablehlo.subtract %[[SUBTRACT_17]], %[[DIVIDE_32]] : tensor<f64>
 // CHECK:           %[[SELECT_12:.*]] = stablehlo.select %[[COMPARE_12]], %[[SUBTRACT_18]], %[[SUBTRACT_17]] : tensor<i1>, tensor<f64>
-// CHECK:           %[[COMPARE_13:.*]] = stablehlo.compare LE, %[[ARG1]], %[[CONSTANT_73]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[COMPARE_13:.*]] = stablehlo.compare EQ, %[[ARG1]], %[[CONSTANT_73]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[COMPARE_14:.*]] = stablehlo.compare LT, %[[ARG1]], %[[CONSTANT_73]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
 // CHECK:           %[[FLOOR_5:.*]] = stablehlo.floor %[[ARG1]] : tensor<f64>
-// CHECK:           %[[COMPARE_14:.*]] = stablehlo.compare EQ, %[[ARG1]], %[[FLOOR_5]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
-// CHECK:           %[[AND_3:.*]] = stablehlo.and %[[COMPARE_13]], %[[COMPARE_14]] : tensor<i1>
+// CHECK:           %[[COMPARE_15:.*]] = stablehlo.compare EQ, %[[ARG1]], %[[FLOOR_5]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[AND_3:.*]] = stablehlo.and %[[COMPARE_14]], %[[COMPARE_15]] : tensor<i1>
 // CHECK:           %[[CONSTANT_96:.*]] = stablehlo.constant dense<0x7FF8000000000000> : tensor<f64>
 // CHECK:           %[[SELECT_13:.*]] = stablehlo.select %[[AND_3]], %[[CONSTANT_96]], %[[SELECT_12]] : tensor<i1>, tensor<f64>
-// CHECK:           %[[SELECT_14:.*]] = stablehlo.select %[[COMPARE_11]], %[[SELECT_13]], %[[MULTIPLY_42]] : tensor<i1>, tensor<f64>
+// CHECK:           %[[CONSTANT_98:.*]] = stablehlo.constant dense<0xFFF0000000000000> : tensor<f64>
+// CHECK:           %[[SELECT_16:.*]] = stablehlo.select %[[COMPARE_13]], %[[CONSTANT_98]], %[[SELECT_13]] : tensor<i1>, tensor<f64>
+// CHECK:           %[[SELECT_14:.*]] = stablehlo.select %[[COMPARE_11]], %[[SELECT_16]], %[[MULTIPLY_42]] : tensor<i1>, tensor<f64>
 // CHECK:           %[[FLOOR_6:.*]] = stablehlo.floor %[[ARG0]] : tensor<f64>
-// CHECK:           %[[COMPARE_15:.*]] = stablehlo.compare NE, %[[ARG0]], %[[FLOOR_6]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
-// CHECK:           %[[COMPARE_16:.*]] = stablehlo.compare LT, %[[ARG0]], %[[CONSTANT_70]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
-// CHECK:           %[[OR_0:.*]] = stablehlo.or %[[COMPARE_15]], %[[COMPARE_16]] : tensor<i1>
+// CHECK:           %[[COMPARE_16:.*]] = stablehlo.compare NE, %[[ARG0]], %[[FLOOR_6]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[COMPARE_17:.*]] = stablehlo.compare LT, %[[ARG0]], %[[CONSTANT_70]] : (tensor<f64>, tensor<f64>) -> tensor<i1>
+// CHECK:           %[[OR_0:.*]] = stablehlo.or %[[COMPARE_16]], %[[COMPARE_17]] : tensor<i1>
 // CHECK:           %[[CONSTANT_97:.*]] = stablehlo.constant dense<0x7FF8000000000000> : tensor<f64>
 // CHECK:           %[[SELECT_15:.*]] = stablehlo.select %[[OR_0]], %[[CONSTANT_97]], %[[SELECT_14]] : tensor<i1>, tensor<f64>
 // CHECK:           return %[[SELECT_15]] : tensor<f64>
@@ -3724,17 +3739,20 @@
 // CHECK:           %[[DIVIDE_32:.*]] = stablehlo.divide %[[MULTIPLY_52]], %[[SINE_1]] : tensor<f32>
 // CHECK:           %[[SUBTRACT_18:.*]] = stablehlo.subtract %[[SUBTRACT_17]], %[[DIVIDE_32]] : tensor<f32>
 // CHECK:           %[[SELECT_12:.*]] = stablehlo.select %[[COMPARE_12]], %[[SUBTRACT_18]], %[[SUBTRACT_17]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[COMPARE_13:.*]] = stablehlo.compare LE, %[[CONVERT_1]], %[[CONSTANT_73]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_13:.*]] = stablehlo.compare EQ, %[[CONVERT_1]], %[[CONSTANT_73]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_14:.*]] = stablehlo.compare LT, %[[CONVERT_1]], %[[CONSTANT_73]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
 // CHECK:           %[[FLOOR_5:.*]] = stablehlo.floor %[[CONVERT_1]] : tensor<f32>
-// CHECK:           %[[COMPARE_14:.*]] = stablehlo.compare EQ, %[[CONVERT_1]], %[[FLOOR_5]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[AND_3:.*]] = stablehlo.and %[[COMPARE_13]], %[[COMPARE_14]] : tensor<i1>
+// CHECK:           %[[COMPARE_15:.*]] = stablehlo.compare EQ, %[[CONVERT_1]], %[[FLOOR_5]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[AND_3:.*]] = stablehlo.and %[[COMPARE_14]], %[[COMPARE_15]] : tensor<i1>
 // CHECK:           %[[CONSTANT_96:.*]] = stablehlo.constant dense<0x7FC00000> : tensor<f32>
 // CHECK:           %[[SELECT_13:.*]] = stablehlo.select %[[AND_3]], %[[CONSTANT_96]], %[[SELECT_12]] : tensor<i1>, tensor<f32>
-// CHECK:           %[[SELECT_14:.*]] = stablehlo.select %[[COMPARE_11]], %[[SELECT_13]], %[[MULTIPLY_42]] : tensor<i1>, tensor<f32>
+// CHECK:           %[[CONSTANT_98:.*]] = stablehlo.constant dense<0xFF800000> : tensor<f32>
+// CHECK:           %[[SELECT_16:.*]] = stablehlo.select %[[COMPARE_13]], %[[CONSTANT_98]], %[[SELECT_13]] : tensor<i1>, tensor<f32>
+// CHECK:           %[[SELECT_14:.*]] = stablehlo.select %[[COMPARE_11]], %[[SELECT_16]], %[[MULTIPLY_42]] : tensor<i1>, tensor<f32>
 // CHECK:           %[[FLOOR_6:.*]] = stablehlo.floor %[[CONVERT_0]] : tensor<f32>
-// CHECK:           %[[COMPARE_15:.*]] = stablehlo.compare NE, %[[CONVERT_0]], %[[FLOOR_6]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[COMPARE_16:.*]] = stablehlo.compare LT, %[[CONVERT_0]], %[[CONSTANT_70]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
-// CHECK:           %[[OR_0:.*]] = stablehlo.or %[[COMPARE_15]], %[[COMPARE_16]] : tensor<i1>
+// CHECK:           %[[COMPARE_16:.*]] = stablehlo.compare NE, %[[CONVERT_0]], %[[FLOOR_6]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[COMPARE_17:.*]] = stablehlo.compare LT, %[[CONVERT_0]], %[[CONSTANT_70]] : (tensor<f32>, tensor<f32>) -> tensor<i1>
+// CHECK:           %[[OR_0:.*]] = stablehlo.or %[[COMPARE_16]], %[[COMPARE_17]] : tensor<i1>
 // CHECK:           %[[CONSTANT_97:.*]] = stablehlo.constant dense<0x7FC00000> : tensor<f32>
 // CHECK:           %[[SELECT_15:.*]] = stablehlo.select %[[OR_0]], %[[CONSTANT_97]], %[[SELECT_14]] : tensor<i1>, tensor<f32>
 // CHECK:           %[[CONVERT_2:.*]] = stablehlo.convert %[[SELECT_15]] : (tensor<f32>) -> tensor<f16>
diff --ruN a/stablehlo/stablehlo/transforms/ChloLegalizeToStablehlo.cpp b/stablehlo/stablehlo/transforms/ChloLegalizeToStablehlo.cpp
--- stablehlo/stablehlo/transforms/ChloLegalizeToStablehlo.cpp
+++ stablehlo/stablehlo/transforms/ChloLegalizeToStablehlo.cpp
@@ -1668,18 +1668,27 @@
   digamma = mlir::stablehlo::SelectOp::create(rewriter, loc, needToReflect,
                                               reflection, digamma);
 
-  // Digamma has poles at negative integers and zero; return nan for those.
-  Value isLeZero = mlir::stablehlo::CompareOp::create(
-      rewriter, loc, x, zero, mlir::stablehlo::ComparisonDirection::LE);
+  // Digamma has poles at negative integers and zero; return nan for
+  // negative integers, and -inf for zero.
+  Value isZero = mlir::stablehlo::CompareOp::create(
+      rewriter, loc, x, zero, mlir::stablehlo::ComparisonDirection::EQ);
+  Value isLtZero = mlir::stablehlo::CompareOp::create(
+      rewriter, loc, x, zero, mlir::stablehlo::ComparisonDirection::LT);
   Value isInt = mlir::stablehlo::CompareOp::create(
       rewriter, loc, x, mlir::stablehlo::FloorOp::create(rewriter, loc, x),
       mlir::stablehlo::ComparisonDirection::EQ);
-  Value isPole = mlir::stablehlo::AndOp::create(rewriter, loc, isLeZero, isInt);
-  return mlir::stablehlo::SelectOp::create(
-      rewriter, loc, isPole,
+  Value isNegativeInteger =
+      mlir::stablehlo::AndOp::create(rewriter, loc, isLtZero, isInt);
+  Value resultWithNan = mlir::stablehlo::SelectOp::create(
+      rewriter, loc, isNegativeInteger,
       getConstantLike(rewriter, loc, std::numeric_limits<double>::quiet_NaN(),
                       x),
       digamma);
+  return mlir::stablehlo::SelectOp::create(
+      rewriter, loc, isZero,
+      getConstantLike(rewriter, loc, -std::numeric_limits<double>::infinity(),
+                      x),
+      resultWithNan);
 }
 
 namespace {

